lib/Sema/CSSolver.cpp

```diff
@@ -537,199 +512,107 @@ ConstraintSystem::SolverScope::~SolverScope() {
 /// \brief Try each of the given type variable bindings to find solutions
 /// to the given constraint system.
 ///
-/// \param depth The depth of the solution stack.
 /// \param typeVar The type variable we're binding.
-/// \param bindings The initial set of bindings to explore.
+/// \param initialBindings The initial set of bindings to explore.
 /// \param solutions The set of solutions.
 ///
 /// \returns true if there are no solutions.
 bool ConstraintSystem::tryTypeVariableBindings(
-    unsigned depth, TypeVariableType *typeVar,
-    ArrayRef<ConstraintSystem::PotentialBinding> bindings,
+    TypeVariableType *typeVar,
+    ArrayRef<ConstraintSystem::PotentialBinding> initialBindings,
     SmallVectorImpl<Solution> &solutions,
     FreeTypeVariableBinding allowFreeTypeVariables) {
+  auto &TC = getTypeChecker();
   bool anySolved = false;
-  llvm::SmallPtrSet<CanType, 4> exploredTypes;
-  llvm::SmallPtrSet<TypeBase *, 4> boundTypes;
+  bool sawFirstLiteralConstraint = false;
+
+  auto attemptTypeVarBinding = [&](PotentialBinding &binding) -> bool {
+    auto type = binding.BindingType;
+
+    // Try to solve the system with typeVar := type
+    ConstraintSystem::SolverScope scope(*this);
+    if (binding.DefaultedProtocol) {
+      type = openUnboundGenericType(type, typeVar->getImpl().getLocator());
+      type = type->reconstituteSugar(/*recursive=*/false);
+    } else if (binding.BindingSource == ConstraintKind::ArgumentConversion &&
+               !type->hasTypeVariable() && isCollectionType(type)) {
+      // If the type binding comes from the argument conversion, let's
+      // instead of binding collection types directly let's try to
```

Comment:
Grammar here is a bit wrong, "let's" appears twice

Reply:
Good catch, I'll fix that